File: algorithms/[70]爬楼梯/solution.ts
// topics = ["动态规划"]

function climbStairs(n: number): number {
  // 滚动数组思想优化空间复杂度
  let p = 0;
  let q = 0;
  let r = 1;

  for (let i = 1; i < n + 1; i += 1) {
    p = q;
    q = r;
    r = p + q;
  }

  return r;
}

File: algorithms/[70]爬楼梯/solution.py
# topics = ["动态规划"]

from functools import lru_cache


class Solution:
    def climbStairs(self, n: int) -> int:
        """
        动态规划
        """
        dp = [1] * (n + 1)
        for i in range(2, n + 1):
            dp[i] = dp[i - 2] + dp[i - 1]
        return dp[n]

    @lru_cache(None)
    def climbStairs2(self, n: int) -> int:
        """
        记忆化递归
        """
        if n <= 3:
            return n
        return self.climbStairs2(n - 2) + self.climbStairs2(n - 1)
